### PR TITLE
vm: refactor out logic of auto-registration and external registration

### PR DESCRIPTION
**Pull request by yskopets**
**Please provide a description of this PR:**

Follow-up to https://github.com/istio/istio/pull/44712.
Follow-up to https://github.com/istio/istio/pull/44827.

Refactors out smaller sub-components out of `pilot/pkg/autoregistration/controller.go`:
* logic of connect/disconnect/cleanup lifecycle and `WorkloadEntry` updates - `pilot/pkg/autoregistration`
* logic of auto-registration - `pilot/pkg/autoregistration/internal/autoregistration`
* logic of externally registered workloads - `pilot/pkg/autoregistration/internal/externalregistration`

This PR is merely a refactoring:
  * it doesn't add any new features
  * all unit and e2e tests remain unchanged 

Implementation notes:
* ~~in this PR, the package is still called `pilot/pkg/autoregistration`; the package will be renamed into `pilot/pkg/workloadentry` in a follow-up PR~~
* based on the feedback to https://github.com/istio/istio/pull/44786 (which was the first version of this PR),  the logic of `CreateWorkloadEntry(...)`, `ChangeWorkloadEntryStateToConnected(...)`, `ChangeWorkloadEntryStateToDisconnected(...)` stays in `pilot/pkg/autoregistration/controller.go`

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -327,8 +327,8 @@ type Proxy struct {
 	// XdsNode is the xDS node identifier
 	XdsNode *core.Node
 
-	WorkloadEntryName        string
-	WorkloadEntryAutoCreated bool
+	WorkloadEntryName           string
+	WorkloadEntryAutoRegistered bool
```
Comment (yskopets) [inline pilot/pkg/model/context.go:331]: `AutoCreated` is renamed into `AutoRegistered` based on [feedback](https://github.com/istio/istio/pull/44712#pullrequestreview-1414150727) in https://github.com/istio/istio/pull/44712

File: pilot/pkg/autoregistration/controller.go [hunk 2]
```diff
@@ -578,104 +499,6 @@ func (c *Controller) IsControllerOf(wle *config.Config) bool {
 	return wle.Annotations[WorkloadControllerAnnotation] == c.instanceID
 }
 
-func autoregisteredWorkloadEntryName(proxy *model.Proxy) string {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:581]: moved to `internal/autoregistration/controller.go` "as is"

File: pilot/pkg/autoregistration/controller.go [hunk 3]
```diff
@@ -685,11 +508,3 @@ func makeProxyKey(proxy *model.Proxy) string {
 	}, "~")
 	return key
 }
-
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:688]: moved to `internal/autoregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 4]
```diff
@@ -685,11 +508,3 @@ func makeProxyKey(proxy *model.Proxy) string {
...
-func isAutoRegisteredWorkloadEntry(wle *config.Config) bool {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:689]: moved to `internal/autoregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 5]
```diff
@@ -685,11 +508,3 @@ func makeProxyKey(proxy *model.Proxy) string {
...
-	return wle != nil && wle.Annotations[AutoRegistrationGroupAnnotation] != ""
-}
-
-func isHealthCheckedWorkloadEntry(wle *config.Config) bool {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:693]: moved to `internal/externalregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 6]
```diff
@@ -211,112 +206,62 @@ func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) err
...
 	}
 	return err
 }
 
-// onWorkloadConnect creates/updates WorkloadEntry of the connecting workload.
-//
-// If workload is using auto-registration, WorkloadEntry will be created automatically.
-//
-// If workload is not using auto-registration, WorkloadEntry must already exist.
-func (c *Controller) onWorkloadConnect(entryName string, proxy *model.Proxy, conTime time.Time, autoCreate bool) error {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:252]: replaced by `getRegistrationStrategy(...)`

File: pilot/pkg/autoregistration/controller.go [hunk 7]
```diff
@@ -523,51 +474,21 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
...
-	if err := c.store.Delete(gvk.WorkloadEntry, wle.Name, wle.Namespace, &wle.ResourceVersion); err != nil && !errors.IsNotFound(err) {
-		log.Warnf("failed cleaning up auto-registered WorkloadEntry %s/%s: %v", wle.Namespace, wle.Name, err)
-		autoRegistrationErrors.Increment()
-		return
-	}
-	autoRegistrationDeletes.Increment()
-	log.Infof("cleaned up auto-registered WorkloadEntry %s/%s periodic:%v", wle.Namespace, wle.Name, periodic)
-}
-
-// deleteHealthCondition updates WorkloadEntry of a workload that is not using auto-registration
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:562]: moved to `internal/externalregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 8]
```diff
@@ -523,51 +474,21 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
...
-// to remove information about the health status (since we can no longer be certain about it).
-func (c *Controller) deleteHealthCondition(wle config.Config, periodic bool) {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:564]: moved to `internal/externalregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 9]
```diff
@@ -523,51 +474,21 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
...
-	}
-	if isHealthCheckedWorkloadEntry(&wle) && health.HasHealthCondition(&wle) {
-		c.deleteHealthCondition(wle, periodic)
-		return
-	}
-}
-
-// deleteEntry removes WorkloadEntry that was created automatically for a workload
-// that is using auto-registration.
-func (c *Controller) deleteEntry(wle config.Config, periodic bool) {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:552]: moved to `internal/autoregistration/controller.go`

File: pilot/pkg/autoregistration/controller.go [hunk 10]
```diff
@@ -523,51 +474,21 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
...
+	if err == nil && time.Since(disconnAt) < cleanupGracePeriod {
 		return false
 	}
 
 	return true
 }
 
-// cleanupEntry performs clean-up actions on a WorkloadEntry of a proxy that hasn't
-// reconnected within a grace period.
-func (c *Controller) cleanupEntry(wle config.Config, periodic bool) {
```
Comment (yskopets) [inline pilot/pkg/autoregistration/controller.go:535]: replaced by `newCleanupTask(...)`

File: pilot/pkg/autoregistration/controller.go [hunk 11]
```diff
@@ -23,67 +23,27 @@ import (
 	"time"
```
Comment (stevenctl) [inline pilot/pkg/autoregistration/controller.go:23]: nit: this is no longer the "autoregistration" controller, it's the WorkloadEntry controller

**Review by stevenctl — COMMENTED** [review]
LGTM. Left a hold if you want to change the top-level package name. 

I also think it's not  super necessary to split everything into internal sub-packages, but not that concerned about it.

File: pilot/pkg/autoregistration/controller.go (hunk 11 above)
Comment (yskopets) [reply pilot/pkg/autoregistration/controller.go:23]: yes, I'm planning to rename the package into `workloadentry`.

I left such renaming for a separate PR to improve readability of this one.

**Comment by yskopets** [comment]
I've rename package `pilot/pkg/autoregistration` into `pilot/pkg/workloadentry`.

**Comment by yskopets** [comment]
> I also think it's not super necessary to split everything into internal sub-packages, but not that concerned about it.

I'm not insist on merging this PR. If you think it would be better without it, I can close it.

**Comment by hzxuzhonghu** [comment]
If it is not that bad, i would like to keep as it was. Refactoring is to keep it simple to maintain. But keep in mind it also makes it hard to track the history.  Sometimes we always forget why we did like this not that, in that case we need to check by git blame

**Pull request closed without merging**